.github/workflows/tag.yml

```diff
@@ -31,11 +32,54 @@ jobs:
         id: create_tag
         uses: mikefarah/yq@v4
         with:
-          cmd: yq -r '"v" + (.MAJ | tostring) + "." + (.MIN | tostring)' ${{ steps.workflow_directory.outputs.name }}/VERSION.yaml
+          cmd: yq -r '"v" + (.MAJ | tostring) + "." + (.MIN | tostring) + "." + (.PATCH | tostring)' ${{ steps.workflow_directory.outputs.name }}/VERSION.yaml
+      - name: Validate tag doesn't already exist
+        run: |
+          if git tag -l | grep -q "${{ steps.create_tag.outputs.result }}"; then
+            echo "Tag ${{ steps.create_tag.outputs.result }} already exists. Exiting."
+            exit 1
+          else
+            echo "Tag ${{ steps.create_tag.outputs.result }} does not exist. Proceeding."
+          fi
+      - name: Validate tag contains all three semantic elements
+        run: |
+          if [[ "${{ steps.create_tag.outputs.result }}" =~ ^v[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
+            echo "Tag ${{ steps.create_tag.outputs.result }} is valid, because it adheres to vMAJ.MIN.PATCH format."
+          else
+            echo "Tag ${{ steps.create_tag.outputs.result }} is not valid. It must adhere to vMAJ.MIN.PATCH format."
+            exit 1
+          fi
+      - name: If this is a PR to a patch branch, ensure the patch version is not 0
+        if: startsWith(github.base_ref, 'patch-')
+        run: |
+          patch_version=$(echo "${{ steps.create_tag.outputs.result }}" | cut -d '.' -f 3)
+          if [ "$patch_version" -ne 0 ]; then
+            echo "Patch version is not 0. This is an acceptable PR to a patch branch."
+          else
+            echo "Patch version is 0. Please ensure that the patch version is incremented before merging this PR."
+            exit 1
+          fi
+
+  cut-release:
+    needs: generate-tag
+    # only actually cut a tag + release when PRs to main or patch branches are merged
+    if: github.event.action == 'closed' && github.event.pull_request.merged == true
+    runs-on: ubuntu-latest
+    permissions:
+      contents: write
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ github.event.pull_request.merge_commit_sha }}
+          fetch-depth: 0  # get existing tags
```

Comment:
I suppose this would _probably_ work without an explicit `ref` argument, but using the [`merge_commit_sha`](https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#:~:text=After%20merging%20a,was%20updated%20to.) somehow felt more explicit to me.